File: src/main/java/algo/prefixSum/MaximumNumberOfNonOverlappingSubarraysWithSumEqualsTarget_1546.java
package algo.prefixSum;

import java.util.HashSet;

public class MaximumNumberOfNonOverlappingSubarraysWithSumEqualsTarget_1546 {
  public static int maxNonOverlapping(int[] nums, int target) {
    var prefixSum = new HashSet<Integer>();

    int count = 0, sum = 0;
    for (int num : nums) {
      sum += num;
      if (sum == target || prefixSum.contains(sum - target)) {
        count++;
        sum = 0;
        prefixSum.clear();
      }
      prefixSum.add(sum);
    }

    return count;
  }

  public static void main(String[] args) {
    System.out.println(maxNonOverlapping(new int[]{1,1,1,1,1}, 2));
    System.out.println(maxNonOverlapping(new int[]{-1,3,5,1,4,2,-9}, 6));
  }
}

File: src/main/kotlin/algo/prefixSum/MaximumNumberOfNonOverlappingSubarraysWithSumEqualsTarget_1546.kt
package algo.prefixSum

class KtMaximumNumberOfNonOverlappingSubarraysWithSumEqualsTarget_1546 {
    fun maxNonOverlapping(nums: IntArray, target: Int): Int {
        val prefixSum: MutableSet<Int> = mutableSetOf()
        var count = 0
        var sum = 0
        nums.indices.forEach { i ->
            sum += nums[i]
            if (sum == target || prefixSum.contains(sum - target)) {
                count++
                sum = 0
                prefixSum.clear()
            }
            prefixSum.add(sum)
        }
        return count
    }
}

fun main() {
    val subarray = KtMaximumNumberOfNonOverlappingSubarraysWithSumEqualsTarget_1546()
    println(subarray.maxNonOverlapping(intArrayOf(1,1,1,1,1), 2))
    println(subarray.maxNonOverlapping(intArrayOf(-1,3,5,1,4,2,-9), 6))
}